# Task 1 – On the use of distance information for UwaveGesture Recognition Task


In [11]:
#import required libraries
import pandas as pd
import matplotlib.pyplot as plt
import numpy as np
import seaborn as sns         
from math import sqrt
from sklearn.neighbors import KNeighborsClassifier
from sklearn.model_selection import cross_val_score
from sklearn.metrics import confusion_matrix,accuracy_score
from sklearn import metrics
from datetime import datetime
from sklearn.linear_model import LogisticRegressionCV
import seaborn as sn
from sklearn.metrics.pairwise import euclidean_distances
from sklearn.linear_model import LogisticRegression
sns.set()

In [2]:
#read train set
X_gesture_train = pd.read_csv('uWaveGestureLibrary_X_TRAIN', sep=" ", prefix="t", header=None, skipinitialspace=True)
Y_gesture_train = pd.read_csv('uWaveGestureLibrary_Y_TRAIN', sep=" ", prefix="t", header=None, skipinitialspace=True)
Z_gesture_train = pd.read_csv('uWaveGestureLibrary_Z_TRAIN', sep=" ", prefix="t", header=None, skipinitialspace=True)
#read test set
X_gesture_test = pd.read_csv('uWaveGestureLibrary_X_TEST', sep=" ", prefix="t", header=None, skipinitialspace=True)
Y_gesture_test = pd.read_csv('uWaveGestureLibrary_Y_TEST', sep=" ", prefix="t", header=None, skipinitialspace=True)
Z_gesture_test = pd.read_csv('uWaveGestureLibrary_Z_TEST', sep=" ", prefix="t", header=None, skipinitialspace=True)

In [3]:
#rename columns name of train set
X_gesture_train.columns = ['X{}'.format(i) for i in range(0,len(X_gesture_train.columns))]
Y_gesture_train.columns = ['Y{}'.format(i) for i in range(0,len(Y_gesture_train.columns))]
Z_gesture_train.columns = ['Z{}'.format(i) for i in range(0,len(Z_gesture_train.columns))]
#merge x, y and z coordinates 
allxyz_train=pd.concat([X_gesture_train.iloc[:,:],Y_gesture_train.iloc[:,1:],Z_gesture_train.iloc[:,1:]], axis=1)
print("merged train set size=", allxyz_train.shape)

#rename columns name of test set
X_gesture_test.columns = ['X{}'.format(i) for i in range(0,len(X_gesture_test.columns))]
Y_gesture_test.columns = ['Y{}'.format(i) for i in range(0,len(Y_gesture_test.columns))]
Z_gesture_test.columns = ['Z{}'.format(i) for i in range(0,len(Z_gesture_test.columns))]
#merge x, y and z coordinates 
allxyz_test=pd.concat([X_gesture_test.iloc[:,:],Y_gesture_test.iloc[:,1:],Z_gesture_test.iloc[:,1:]], axis=1)
print("merged test set size=", allxyz_test.shape)

merged train set size= (896, 946)
merged test set size= (3582, 946)


a) Suppose we decided to apply a nearest-neighbor (NN) classifier to find the labels of test
instances. You can use the strategy you employed when you apply PCA to this data in
Homework 2 (i.e. concatenation of the axes). Propose two distance measures for computing
similarity between two time series. The distance calculation on the concatenated time series
implicitly weights the distances of each axis in an equal way. For each distance measure
alternative, use the training data to identify the ideal value of k which minimizes the error of a
10-fold cross-validation.

b) Using the value of k (identified for each distance measure) in part (a) and evaluate your final
performance on the test data and present your results in a (8-by-8) confusion matrix, showing
the counts for actual and predicted labels. In addition, quote the runtime and accuracy for your
results.

In [4]:
#split the data as input and the output for both train and test set
X_train=allxyz_train.iloc[:,1:]
Y_train=allxyz_train.iloc[:,:1:]
X_test=allxyz_test.iloc[:,1:]
Y_test=allxyz_test.iloc[:,:1:]

In [6]:
#find ideal valu of k for euclidean metric
k_values = [i for i in range(1,50,2)]
k_acc=[]
for i in k_values: 
    knn = KNeighborsClassifier(n_neighbors = i, metric = 'euclidean', n_jobs=-1)
    cv_scores=cross_val_score(knn, X_train, np.ravel(Y_train), cv=10, scoring='accuracy')
    k_acc.append(cv_scores.mean())
optimal_k=k_values[k_acc.index(max(k_acc))]
print("Optimal k for euclidean=", format(optimal_k))

Optimal k for euclidean= 3


In [7]:
#for euclidean the optimum n_neigbors value is 3
#find the confusion matrix and the accuracy for euclidean
start_time = datetime.now()

classifier_e = KNeighborsClassifier(n_neighbors = 3, metric = 'euclidean', p = 2)
classifier_e.fit(X_train, np.ravel(Y_train))
y_pred = classifier_e.predict(X_test)

cm_e = confusion_matrix(Y_test, y_pred)
ac_e = accuracy_score(Y_test,y_pred)
print("Our confusion matrix is \n", pd.DataFrame(cm_e))
print("Our accuracy is", ac_e)

end_time = datetime.now()
print('Duration: {}'.format(end_time - start_time))

Our confusion matrix is 
      0    1    2    3    4    5    6    7
0  431    0    0    2    0    4    0    0
1    1  449    0    0    0    0    2    0
2    2    0  413    0   15   20    4    0
3    7    0    0  370   60    6    0    7
4    3    0    6    1  422    1    0    0
5    6    0    7   15   28  392    1    0
6    0    0    1    0    0    0  446    0
7    0    0    0    1    1    0    0  458
Our accuracy is 0.9438860971524288
Duration: 0:00:04.635774


In [8]:
#find ideal valu of k for manhattan metric
k_values = [i for i in range(1,50,2)]
k_acc=[]
for i in k_values: 
    knn = KNeighborsClassifier(n_neighbors = i, metric = 'manhattan', n_jobs=-1)
    cv_scores=cross_val_score(knn, X_train, np.ravel(Y_train), cv=10, scoring='accuracy')
    k_acc.append(cv_scores.mean())
optimal_k=k_values[k_acc.index(max(k_acc))]
print("Optimal k for manhattan=", format(optimal_k))

Optimal k for manhattan= 1


In [9]:
#for manhattan the optimum n_neigbors value is 1
#find the confusion matrix and the accuracy for manhattan
start_time = datetime.now()

classifier_m = KNeighborsClassifier(n_neighbors = 1, metric = 'manhattan', p = 2)
classifier_m.fit(X_train, np.ravel(Y_train))
y_pred = classifier_m.predict(X_test)

cm_m = confusion_matrix(Y_test, y_pred)
ac_m = accuracy_score(Y_test,y_pred)

print (confusion_matrix)
print("Our confusion matrix is \n", pd.DataFrame(cm_m))
print("Our accuracy is", ac_m)

end_time = datetime.now()
print('Duration: {}'.format(end_time - start_time))

<function confusion_matrix at 0x0000021A59A30D90>
Our confusion matrix is 
      0    1    2    3    4    5    6    7
0  430    0    0    2    0    5    0    0
1    1  451    0    0    0    0    0    0
2    2    0  418    0   16   14    4    0
3    3    0    0  392   40   11    0    4
4    3    0    9    5  415    1    0    0
5    3    0    5   13   16  411    0    1
6    0    0    2    0    0    0  445    0
7    0    0    0    3    1    0    0  456
Our accuracy is 0.954215522054718
Duration: 0:00:04.031079


c) The observations from different axes are weighted equally if we compute the distance over each
axis and sum them to obtain a final similarity measure. Is this reasonable? For example, we can
compute the distance as below:


where DistX is the distance based on the acceleration only on X axis, DistY is for Y axis and so
on. Do you think weighting the distances over different axes to obtain a final similarity measure
makes sense for classification? Why?


C) Yes, it makes sense. Calculating the distances over x and y is more reasonable because the z causes extra information. 2D calculation is enough that is why the weight of z is less effective in this work.

# Task 2 – Linear models on alternative representations of the data

In [13]:
#convert Class 3 to 1 and the others to 0
Y_train3 = (allxyz_train["X0"]==3)*1
Y_test3 = (allxyz_test["X0"]==3)*1

a) Train a logistic regression model on the training data and use the model to make a prediction on the
test data. Note that you will obtain probabilistic predictions (i.e. probability of a time series being from
Class 3 if you encoded Class 3 as 1 in binary classification setting). This will require you to select a
threshold since 0.5 as a threshold may not work well under this imbalanced class setting. To make
things easier, use the ratio of Class 3 instances in the training data as threshold. Use the learned model
to predict the class for test data. Present your results in a (2-by-2) confusion matrix.


In [14]:
#train a logistic regression model on train set
model = LogisticRegression(max_iter=1000)
model.fit(X_train, np.ravel(Y_train3))

C:\Users\90505\Anaconda3\lib\site-packages\sklearn\linear_model\logistic.py:432: FutureWarning: Default solver will be changed to 'lbfgs' in 0.22. Specify a solver to silence this warning.
  FutureWarning)


LogisticRegression(C=1.0, class_weight=None, dual=False, fit_intercept=True,
                   intercept_scaling=1, l1_ratio=None, max_iter=1000,
                   multi_class='warn', n_jobs=None, penalty='l2',
                   random_state=None, solver='warn', tol=0.0001, verbose=0,
                   warm_start=False)

In [15]:
#calculate the treshold
th=Y_train3.sum()/len(X_train)
th=float(th)

In [16]:
#make prediction on test set
probs_y=model.predict_proba(X_test) 
probs_y=(probs_y[:,1]>th)*1

In [17]:
#2-by-2 confusion matrix
conf_matrix = metrics.confusion_matrix(Y_test3, probs_y)
print("Our confusion matrix is \n", pd.DataFrame(conf_matrix))

Our confusion matrix is 
       0    1
0  2828  300
1    59  395


In [19]:
accuracy1=(conf_matrix[0][0]+conf_matrix[1][1])/conf_matrix.sum()
print("Accuracy is= %.3f" % accuracy1)

Accuracy is= 0.900


b) An advantage of logistic regression is related to the interpretability however when we have large
number of features together with a method without penalization, it is harder to interpret the results.
Therefore, an alternative way is to train a logistic regression model with lasso penalties. This will
require you setting of penalization term (namely lambda). Use 10-fold cross-validation to determine
your ideal lambda level based on binomial deviance (Note that we have used accuracy as primary
metric to determine the lambda in class, however this strategy may not work well for the imbalanced
data). You can check http://www.inf.ed.ac.uk/teaching/courses/mlsc/Notes/Lecture4/MLSC_Lec4.pdf
for details of binomial deviance. This is also referred to as logistic loss. If you are using “glmnet”
package in R, “type.measure” can be set to “deviance” which is the default value. If you are Python
user, sklearn module has “LogisticRegressionCV” function in which you can provide the scorer as
“metrics.log_loss”.
Once you determine your best lambda value using 10-fold cross-validation, perform classification on
test data similar to part a and compare your results. Comment on the regression coefficients. Is there
any interesting information? Try to interpret the model.

In [20]:
#train logistic regression model with lasso penalties
model_lasso = LogisticRegressionCV(cv=10, penalty="l1", max_iter=1000, scoring='neg_mean_absolute_error', solver='liblinear')
model_lasso.fit(X_train, np.ravel(Y_train3))
pred_y_lasso = model_lasso.predict_proba(X_test)

In [21]:
#calculate the loss due to the chosen lambda values by the 10-fold cross validation above
pred_y_lasso = (pred_y_lasso[:,1]>th)*1
score = metrics.log_loss(Y_test3, pred_y_lasso)
print("best score=", score) 

best score= 2.863827716750354


In [22]:
#2-by-2 confusion matrix for LR lasso model
conf_matrix_lasso = metrics.confusion_matrix(Y_test3, pred_y_lasso)
print("Our confusion matrix is \n", pd.DataFrame(conf_matrix_lasso))

Our confusion matrix is 
       0    1
0  2860  268
1    29  425


In [127]:
#accuracy calculation
accuracy2=(conf_matrix_lasso[0][0]+conf_matrix_lasso[1][1])/conf_matrix_lasso.sum()
print("Accuracy is= %.3f" % accuracy2)

Accuracy is= 0.917


In [114]:
xyz = pd.DataFrame((model_lasso.coef_))
x1 = xyz.iloc[:,:315]
x = ((x1.to_numpy()!=0)*1).sum()
sumx = x1.to_numpy().sum()
y1 = xyz.iloc[:,315:630]
y = ((y1.to_numpy()!=0)*1).sum()
sumy = y1.to_numpy().sum()
z1 = xyz.iloc[:,630:]
z = ((z1.to_numpy()!=0)*1).sum()
sumz = z1.to_numpy().sum()
print("x = ", x)
print("y = ", y)
print("z = ", z)
print("sumx = ", sumx)
print("sumy = ", sumy)
print("sumz = ", sumz)

x =  35
y =  19
z =  10
sumx =  2.510454815309821
sumy =  1.4170113139440206
sumz =  -0.4088124865978635


When the coefficients are interpreted, it can be seen that X axis is the most effective one by it is possitive coefficients. It has the most non zero values also. X direction is the significant one.

c) An alternative way to represent the feature matrix on a new space to introduce nonlinear relations is
to use distance matrix as a feature matrix. For example, we have 896 training instances and the
observations over time are used as features in the previous tasks (i.e. we worked on N by 3T matrix).
Recall that multidimensional scaling also works on distance matrices and we have mentioned that it can
handle nonlinear relations (Homework 2 also aims at revealing such an information). This non-linearity
stems from the use of Euclidean distances. Use of Euclidean distance as input to a learning algorithm
allows for handling nonlinear relations*. In other words, your features keep the nonlinear information.
*We will have further discussion on this behavior when we cover support vector machines. This type of transformations are
discussed under “distance-based kernels” which is out of scope for now. Additional information is provided in case you are
willing to perform research on distance based transformations.
Given this information, you are expected to transform your training data to distance information (i.e. N
by N matrix). Note that you need to perform a similar transformation to your test data. In other words,
you need to calculate the distance of each test instance to training instance to obtain a distance based
representation for your test data. This will be an Ntest by N matrix (Ntest refers to the number of test
instances) where each entry (i,j) refers to the distance of test time series i to the training time series j.
You can use Euclidean distance as your distance measure.
Perform the same training and test strategy as in part b but use the distances as your new feature
matrices. Comment on the regression coefficients. What do they imply under this new representation
setting? 

In [115]:
#creation of feature matrix by the euclidean distance
train = euclidean_distances(X_train)
train = pd.DataFrame(train)
test = euclidean_distances(X_test, X_train)
test = pd.DataFrame(test)
test.head()

0          1          2          3          4          5    \
0  51.687559  13.263419  28.321555  39.541255  34.286788  38.211025   
1  34.947350  54.334337  54.694516  46.732487  45.800128  46.545913   
2  42.171916  25.570757  35.860929  42.167440  28.760080  37.984422   
3  41.876963  34.269027  41.884193  44.840555  25.517543  43.336345   
4  49.642237  14.399477  16.490769  41.308488  32.737889  37.112181   

         6          7          8          9    ...        886        887  \
0  36.045863  48.764229  37.568076  59.428712  ...  44.366890  49.494730   
1  41.582900  37.517528  46.173264  31.870273  ...  37.760887  42.294340   
2  41.091221  44.311017  31.095472  50.164764  ...  46.952663  50.504509   
3  38.773427  38.422948  20.204116  50.381287  ...  45.197983  53.934857   
4  44.964582  47.155949  37.856534  57.156578  ...  47.806687  46.909740   

         888        889        890        891        892        893  \
0  32.777330  51.240891  35.257192  43.642435  51.738237  41.011374   
1  44.207834  38.800478  50.753346  42.983088  38.362504  44.871331   
2  30.280558  39.453718  45.695170  47.538985  50.386797  40.439901   
3  28.368976  42.632988  48.775380  49.697657  44.197285  42.866915   
4  32.783804  49.979809  34.794183  40.932294  53.618109  46.638917   

         894        895  
0  41.870528  28.349480  
1  37.517618  55.648530  
2  43.688492  35.511051  
3  42.667953  41.614170  
4  47.715370  22.199250  

[5 rows x 896 columns]

In [116]:
#train logistic regression model with lasso penalties
model_lasso_2 = LogisticRegressionCV(cv=10, penalty="l1", max_iter=1000, scoring='neg_mean_absolute_error', solver='liblinear')
model_lasso_2.fit(train, np.ravel(Y_train3))
pred_y_lasso_2 = model_lasso_2.predict_proba(test)

In [117]:
#calculate the loss due to the chosen lambda values by the 10-fold cross validation above
pred_y_lasso_2 = (pred_y_lasso_2[:,1]>th)*1
score = metrics.log_loss(Y_test3, pred_y_lasso_2)
print("best score=", score) 

best score= 0.4917641499919217


In [118]:
#2-by-2 confusion matrix
conf_matrix_lasso_2 = metrics.confusion_matrix(Y_test3, pred_y_lasso_2)
print("Our confusion matrix is \n", pd.DataFrame(conf_matrix_lasso_2))

Our confusion matrix is 
       0    1
0  3101   27
1    24  430


In [128]:
accuracy3=(conf_matrix_lasso_2[0][0]+conf_matrix_lasso_2[1][1])/conf_matrix_lasso_2.sum()
print("Accuracy is= %.3f" % accuracy3)

Accuracy is= 0.986


In [126]:
xyz1 = pd.DataFrame((model_lasso_2.coef_))
xyz = ((xyz1.to_numpy()!=0)*1).sum()
sumxyz = (xyz1.to_numpy()).sum()
print("xyz = ", xyz)
print("sumxyz = ", sumxyz)

xyz =  889
sumxyz =  -0.23553478634436253


There are 889 non zero entries and the number of positive output is much more less. The only decision for the model is to understand the class is 3 or not so the valu of the outpu is not important and the model gives a very good performance.

d) Provide an overall comparison on the results you obtain for each part (over all tasks). You can
compare test accuracy of each alternative method you developed.


In [146]:
print("Accuracies from beginning to the end are =\n%.3f " %accuracy1)
print("%.3f" %accuracy2)
print("%.3f" %accuracy3)

Accuracies from beginning to the end are =
0.900 
0.917
0.986


Logistic regression with lasso is better than the logisctic regression model. By the non linear features and lasso penalization, model becomes much better.